File: Code/3D Localization/code/prepare_view_bins_from_tiff.py
# -*- coding: utf-8 -*-
"""Convert Views/{v}/file_{v}.tif -> suite2p/plane0/data.bin (int16).

Streams one TIFF page at a time (peak RAM ~ one frame). Applies ops yoff/xoff
rigid shifts when present. Skips a view if data.bin already has the expected size
and is not all-zero.
"""
from __future__ import annotations

import gc
from pathlib import Path

import numpy as np
import tifffile

from dataset_config import DATA_ROOT, view_plane_dir, view_tiff_path
from tplfm_utils import load_npy

N_VIEWS = 4


def _frame_to_int16(frame: np.ndarray) -> np.ndarray:
    if frame.dtype == np.uint16:
        return np.clip(frame.astype(np.int32), 0, 32767).astype(np.int16)
    clipped = np.asarray(frame, dtype=np.float32)
    return np.clip(np.rint(clipped), -32768, 32767).astype(np.int16)


def _bin_ok(bin_path: Path, nf: int, ly: int, lx: int) -> bool:
    expect = nf * ly * lx * 2
    if not bin_path.is_file() or bin_path.stat().st_size != expect:
        return False
    raw = np.memmap(str(bin_path), dtype=np.int16, mode="r", shape=(nf, ly, lx))
    try:
        for i in (0, nf // 4, nf // 2, (3 * nf) // 4, nf - 1):
            if int(np.max(np.abs(raw[i]))) != 0:
                return True
    finally:
        del raw
    return False


def convert_one(v: int) -> Path:
    plane = view_plane_dir(v, DATA_ROOT)
    ops = load_npy(plane / "ops.npy", allow_pickle=True).item()
    ly, lx = int(ops["Ly"]), int(ops["Lx"])
    nf = int(ops["nframes"])
    out_bin = plane / "data.bin"

    if _bin_ok(out_bin, nf, ly, lx):
        print("view%d: skip existing %s (%d bytes)" % (v, out_bin, out_bin.stat().st_size))
        return out_bin

    tiff_path = view_tiff_path(v, DATA_ROOT)
    expect = nf * ly * lx * 2

    yoff = ops.get("yoff")
    xoff = ops.get("xoff")
    yoff_a = None if yoff is None else np.asarray(yoff).ravel()
    xoff_a = None if xoff is None else np.asarray(xoff).ravel()

    print("view%d: streaming %s -> %s  (%d x %d x %d)" % (v, tiff_path, out_bin, nf, ly, lx))
    n_nz = 0
    with tifffile.TiffFile(str(tiff_path)) as tif:
        n_pages = len(tif.pages)
        if n_pages < nf:
            raise RuntimeError("TIFF pages %d < nframes %d" % (n_pages, nf))
        tmp = out_bin.with_suffix(".bin.tmp")
        with open(tmp, "wb") as f:
            for i in range(nf):
                frame = np.asarray(tif.pages[i].asarray())
                if frame.shape != (ly, lx):
                    raise ValueError(
                        "TIFF page %d shape %s != ops %dx%d" % (i, frame.shape, ly, lx)
                    )
                if yoff_a is not None and xoff_a is not None and i < yoff_a.size:
                    dy, dx = int(yoff_a[i]), int(xoff_a[i])
                    if dy != 0 or dx != 0:
                        frame = np.roll(frame, shift=(-dy, -dx), axis=(0, 1))
                clipped = _frame_to_int16(frame)
                if np.any(clipped):
                    n_nz += 1
                clipped.tofile(f)
                del frame, clipped
                if (i + 1) % 500 == 0:
                    print("    page %d/%d" % (i + 1, nf))
                    gc.collect()
        got = tmp.stat().st_size
        if got != expect:
            tmp.unlink(missing_ok=True)
            raise RuntimeError("temp data.bin size %d != expect %d" % (got, expect))
        tmp.replace(out_bin)

    if n_nz == 0:
        raise RuntimeError("view %d converted movie is all zeros" % v)
    print(
        "  wrote %s  shape=(%d,%d,%d) dtype=int16  nonzero_frames=%d/%d"
        % (out_bin, nf, ly, lx, n_nz, nf)
    )
    gc.collect()
    return out_bin


def main() -> None:
    print("DATA_ROOT =", DATA_ROOT)
    for v in range(1, N_VIEWS + 1):
        convert_one(v)
    print("Done. All view data.bin ready.")


if __name__ == "__main__":
    main()

File: Code/3D Localization/matlab/prepare_view_bins_from_tiff.m
function prepare_view_bins_from_tiff(cfg)
%PREPARE_VIEW_BINS_FROM_TIFF Stream Views/{v}/file_{v}.tif -> suite2p/plane0/data.bin.
if nargin < 1, cfg = dataset_config(); end
fprintf('DATA_ROOT = %s\n', cfg.dataRoot);
for v = 1:4
    convert_one(v, cfg);
end
fprintf('Done. View data.bin check complete.\n');
end

function convert_one(v, cfg)
plane = view_plane_dir(v, cfg);
s2p = load_suite2p(plane);
ly = s2p.ops.Ly;
lx = s2p.ops.Lx;
nf = s2p.ops.nframes;
outBin = fullfile(plane, 'data.bin');
if bin_is_valid(outBin, nf, ly, lx)
    info = dir(outBin);
    fprintf('view%d: skip existing %s (%d bytes)\n', v, outBin, info.bytes);
    return
end
d = fileparts(fileparts(plane));
tiffPath = fullfile(d, sprintf('file_%d.tif', v));
if ~exist(tiffPath, 'file')
    fprintf('view%d: no TIFF and no valid data.bin (skipped)\n', v);
    return
end
info = imfinfo(tiffPath);
if numel(info) < nf
    error('TIFF pages %d < nframes %d', numel(info), nf);
end
yoff = s2p.ops.yoff(:);
xoff = s2p.ops.xoff(:);
fprintf('view%d: streaming %s -> %s  (%d x %d x %d)\n', v, tiffPath, outBin, nf, ly, lx);
tmp = [outBin, '.tmp'];
fid = fopen(tmp, 'wb');
if fid < 0
    error('cannot write %s', tmp);
end
nNz = 0;
try
    for i = 1:nf
        frame = imread(tiffPath, i, 'Info', info);
        if ndims(frame) > 2
            frame = frame(:, :, 1);
        end
        if ~isequal(size(frame), [ly, lx])
            error('TIFF page %d shape mismatch', i);
        end
        if ~isempty(yoff) && ~isempty(xoff) && i <= numel(yoff)
            dy = round(yoff(i));
            dx = round(xoff(i));
            if dy ~= 0 || dx ~= 0
                frame = circshift(frame, [-dy, -dx]);
            end
        end
        clipped = frame_to_int16(frame);
        if any(clipped(:) ~= 0)
            nNz = nNz + 1;
        end
        fwrite(fid, clipped', 'int16'); % C-order X-fastest
        if mod(i, 500) == 0
            fprintf('    page %d/%d\n', i, nf);
        end
    end
catch err
    fclose(fid);
    if exist(tmp, 'file'), delete(tmp); end
    rethrow(err);
end
fclose(fid);
expect = nf * ly * lx * 2;
got = dir(tmp);
if got.bytes ~= expect
    delete(tmp);
    error('temp data.bin size %d != expect %d', got.bytes, expect);
end
if exist(outBin, 'file')
    delete(outBin);
end
movefile(tmp, outBin);
if nNz == 0
    error('view %d converted movie is all zeros', v);
end
fprintf('  wrote %s  shape=(%d,%d,%d) dtype=int16  nonzero_frames=%d/%d\n', ...
    outBin, nf, ly, lx, nNz, nf);
end

function clipped = frame_to_int16(frame)
if isa(frame, 'uint16')
    clipped = int16(min(max(int32(frame), 0), 32767));
else
    clipped = int16(min(max(round(double(frame)), -32768), 32767));
end
end
